### PR TITLE
Contour2d

### PR DESCRIPTION
**Pull request by SoykaDaimler**
Hi all,
based on our discussion in #15 I created a new branch which just includes the suggestions for a 2D contour. And two small additions to .gitignore.
@AndreasVIRES Does this definition comply with the OpenDrive specification?
@nHirsenkorn I added the contour to BaseStationary and BaseMoving, since you did not want it to be mixed up with the Dimension3D bounding box. I hope this fits better?
Best,
Florian

### CODE REVIEW

File: osi_common.proto [hunk 1]
```diff
@@ -154,6 +154,9 @@ message BaseStationary
 
     /// The relative orientation of the landmark w.r.t its parent frame.
     optional Orientation3d orientation = 3;
+    
+    /// The two dimensional (flat) contour of the object.
```
Comment (nHirsenkorn) [inline osi_common.proto]: please further extend the comment: is the last point equal to the first point?

Comment (SoykaDaimler) [inline osi_common.proto]: Good point! I will further clarify. 
I would suggest that the first and last are not equal to avoid redundancy.
Furthermore I would suggest counterclockwise definition of the points. 
Then there is the question about the reference frame. We usually had the middle of the rear axle. But as far as I understood it, in OSI the standard frame is the middle of the bounding box.
@nHirsenkorn Is this correct?

Comment (AndreasVIRES) [inline osi_common.proto]: > Furthermore I would suggest counterclockwise definition of the points.

In OpenDRIVE there is no order foreseen, i.e. the tool would need to be able to calculate and swap the sequence. Is this needed by the sensor modellers?

> Then there is the question about the reference frame.

The contour is given in _relative_ coordinates to that reference point, isn't it?

Comment (nHirsenkorn) [inline osi_common.proto]: @SoykaDaimler I agree to your further definition. Consider adding counterclockwise when in bird-view. The question about reference frames will be clarified in #39

Comment (SoykaDaimler) [inline osi_common.proto]: Our sensor models require counter clockwise ordering, that's why I suggested it. 
I guess in the end the sensor model has to check the order anyway and potentially correct it if the wrong order was provided. So I would keep the requirement for CCW ordering in the interface if no one objects, but keep it up to the model to really check if this requirement was fulfilled.
I agree with the two reference frames mentioned in #39, ground truth should be in world coordinates and the sensed contour in sensor coordinates. We do need access to the specifications where the sensor is mounted to do further transformations ourselves. I guess this is given in relation to the middle of the bounding box? Then we also need the relations between the middle of the bounding box and the middle of the rear/front axis since we have other coordinate systems which are in relation to these points. 
@nHirsenkorn These specs are not yet implemented in OSI, are they?

Comment (AndreasVIRES) [inline osi_common.proto]: > but keep it up to the model to really check if this requirement was fulfilled.

Either it has to be fulfilled by the tool or not. Shouldn't be too complicated for us, I just wanted to clarify.

**Reference frame:**
At the moment I cannot really follow the discussion. Usually you have a position (world, relative, road, whatever), an orientation and a contour (or bounding box) _relative_ to the given position and orientation.

E.g.: 
Object position (reference point): x=10, y=20, z=0
Object orientation: hdg=30°, pit=0°,rol=0°
Object contour: p1[2;0], p2[-2;-2];p3[-2;2]

This results in a rectangle at x=10, y=20, rotated by 30°:
example

Comment (nHirsenkorn) [inline osi_common.proto]: Nope. This is already implemented:
determining the absolute position of a sensor requires two transformations. 
See SensorData::mounting_position and Vehicle::relative_ego_origin

Comment (SoykaDaimler) [inline osi_common.proto]: @AndreasVIRES Thanks for the nice visualization! In our group, we define the contour in world and sensor coordinates, and not relative to the position/orientation.
@nHirsenkorn Ok, thanks for clarifying. Then I will also add the offset vectors for the rear / front axle in the vehicle description.

Comment (AndreasVIRES) [inline osi_common.proto]: @SoykaDaimler This information would also be available with a "position+orientation+contour" definition, you just need to calculate it. Sure, the tool could provide the oultine of an object in world coordinates or sensor coordinates but that is a redundant information because you still need "(reference point)position+orientation".

I also would suggest to discuss the reference point of vehicles again. In my opinion it makes no sense to have different reference point definitions, depending on whether a vehicle is a "Ego" or not. OpenSCENARIO until now does not provide any means to distinguish between vehicles that are an "Ego" and other vehicles. Reference point of a vehicle is always center of rear axis on ground in neutral load conditions.
@raueATconti the whole coordinate topic might be worth discussing it on the workshop.

Comment (nHirsenkorn) [inline osi_common.proto]: @AndreasVIRES All objects have the same reference point. Independent of ego or not. What is causing confusion here is:
A position vector consists of two points:
start point: This is defined by the reference frame. (in OSI: sensor-frame or world frame)
end point: often referred to as reference point. It is **always** the middle of the bounding box. This is not depending on being an ego vehicle or not, and not depending on the coordinate system (reference frame).

Moreover a vector always is dependent of the direction in which it is regarded. This is also captured by the reference frame.

In other words: inconsistencies between ego and other vehicles do not exist.

I will capture this in a wiki article and refer to it here.

Comment (nHirsenkorn) [inline osi_common.proto]: [Coordinate System Wiki page](https://github.com/OpenSimulationInterface/open-simulation-interface/wiki/Coordinate-systems-and-reference-points)

Comment (SoykaDaimler) [inline osi_common.proto]: I think the sensed contour should be given directly in sensor coordinates. This is more natural for a sensor, because a sensor does not sense a position in the middle of a bounding box or where ever, but it senses the contour of an object in sensor coordinates. 
I agree with @AndreasVIRES that we should discuss the coordinate topic in general all together during the workshop.

Comment (AndreasVIRES) [inline osi_common.proto]: @nHirsenkorn I still see an inconsistency because (as the name of the vector indicates) only Ego vehicles provide a vector **relative_ego_origin**, i.e. for non Ego vehicles there is less information on the interface.

In my opinion we need to agree on a vehicle reference point for all vehicles in the simulation, which not necessarily needs to be the rear axis but also not necessarily needs to be the center of the bounding box.

By the way, the real bounding box changes, even during runtime (opening doors but also vehicle movement, spring compression, etc.).

Comment (nHirsenkorn) [inline osi_common.proto]: @AndreasVIRES I see your point: you would like all vehicles to have the same informations. However for non-ego vehicles this information is never required. We should rather consider dropping relative_ego_origin generally. It seems to cause irritations and can in principal be handled by the environment model. I will open an issue.

Comment (nHirsenkorn) [inline osi_common.proto]: @AndreasVIRES the reference point is already identical for all vehicles in the simulation. It never is rear axis, but always the middle of the bounding box.

You are right about the changing bounding box. however it practically has no impact on the sensors. Therefore treating it in a static manner is sufficient :)

Comment (CarloVanDriestenBMW) [inline osi_common.proto]: You guys just tell me when to hit merge alright?  ^^

Comment (nHirsenkorn) [inline osi_common.proto]: nope. We could hit it ourselves xD

Comment (AndreasVIRES) [inline osi_common.proto]: @nHirsenkorn The vehicle reference point is never rear axis?

This is a definition found together with OEMs over years and also the result of many PEGASUS workshops:

> Ego-Fahrzeug:
> - Position: x[m], y[m], z[m] (Weltkoordinaten; **Bezugspunkt Fahrzeug: Mitte Hinterachse**));

This is a blocking point for VIRES for supporting OSI or not and I'll end discussion here, because nobody of the other tool suppliers, modelers and not all OEMs are participating the discussion.

@raueATconti We need to clarify this before we'll start implementation on our side.

Comment (nHirsenkorn) [inline osi_common.proto]: @AndreasVIRES End-Points of Positions always refer to the middle of the bounding box. Finding out the position is solely possible if relative_ego_origin is set to the rear axis. However relative_ego_origin is solely required for transforming from world to sensor-coordinates or for transforming from sensor-coordinates to the environment model. It does not play any role when staying in the same coordinate system (world or sensor).

We explicitly did not choose the rear axis as reference point, as it plays no role for sensors and just increases the complexity of finding out other points on the vehicle.

I get your point, we should end the discussion here. Further commitment would be helpful.
I am also not willing to explain every reasoning multiple times just due to the lack of participation of others.

Comment (SoykaDaimler) [inline osi_common.proto]: I agree and I think that it is very important that the reasoning you (OSI developers) had in mind when creating OSI, is written down somewhere and explained to all the partners who should be using OSI. We need all the PEGASUS partners on board in order to get through this important initial phase of establishing a common interface.

Comment (RudolfVocilka) [inline osi_common.proto]: @SoykaDaimler
> we define the contour in world and sensor coordinates, and not relative to the position/orientation.
If the contour is defined in world coordinates, why is there no Z-component in the contour? On what plane is the z-component projected? Or am I missing something. If z is simply set to 0 there may be problems with lanes with elevation. Or can I imagine it as a list of points on a plane, that is then translated and rotated.

If it is defined in sensor coordinates it should be relative to position / orientation right? 

Regarding the reference frame discussions I would leave it as is for this pull request.

Comment (SoykaDaimler) [inline osi_common.proto]: @RudolfVocilka 
Hi, I just discussed this issue with @CarloVanDriestenBMW on the phone. We use the contour in a purely two dimensional context. That's why I also defined a vector2d in order to avoid confusion about the z coordinate. In a 2D world there is no elevation.

I agree with solving the reference frame discussion elsewhere.

**Comment by SoykaDaimler** [comment]
I updated the description of the contour fields based on our discussion. 
I added offset vectors from the bounding box center point to the middle of the front/rear axle to allow for coordinate transformations. 
Please review these changes.
I think the coordinate system discussion in general has to be solved elsewhere including everybody. 
The changes I suggest can later be adapted to the coordinate system of our choosing.

**Comment by AndreasVIRES** [comment]
1. Does "projection to the street" mean that street parameter like superelevation, etc. are considered? If not maybe we should simply call it "projection to the z-plane".
2. "middle of axis" is not specific enough. What does the z-coordinate of that vector mean?

**Comment by SoykaDaimler** [comment]
Thanks @AndreasVIRES for your questions. 
I clarified the "middle of the axis" description: As I understood it, the center of the reference frame is located in the middle of the axis with respect to all three coordinates x,y and z. So that should define the z component as well.
About the "projection to the street": I rewrote the whole description and definition. I understand the contour as an extension to the bounding box. In ground truth, the contour should be defined once in the beginning in the local object frame and should not change anymore (same as the bounding box). During runtime, one can compute the current state based on the position and orientation data. 
As sensor data, the polygon should describe the visible part of the contour.
So now everything is described in the local object frame and I clarified the difference between ground truth and sensor data. I hope this makes it more clear.

**Comment by AndreasVIRES** [comment]
**Middle of axis** 
Sure, middle in x, y and z but under what circumstances? Usually the definition is "neutral load conditions" or similar. The axles are moving during simulation relatively to the rest of the vehicle, shall your vector reflect this behavior?

**Comment by SoykaDaimler** [comment]
Ah yes, thanks! "Neutral load conditions" is what I meant. I'll update the file...

File: osi_object.proto [hunk 2]
```diff
@@ -54,6 +54,15 @@ message Vehicle
...
     optional Identifier driver = 10;
+    
+    /// The vector pointing from the bounding box center point (base.position) to the middle (in x, y and z) 
+    /// of the rear axle under neutral load conditions.
+    optional Vector3d bbcenter_to_rear = 11;
+
+	/// The vector pointing from the bounding box center point (base.position) to the middle (in x, y and z) 
+	/// of the front axle under neutral load conditions.    		
+	optional Vector3d bbcenter_to_front = 12;
+
```
Comment (CarloVanDriestenBMW) [inline osi_object.proto]: > optional Vector3d relative_ego_origin = 6;

@SoykaDaimler and @AndreasVIRES 
The description above states, that this is the vector to the rear axle, if the car is a simulated car. This is now duplicated? We could consider removing it and replacing it by the two new definitions?

Comment (SoykaDaimler) [inline osi_object.proto]: I agree. In general we need these offsets, no matter if it is the ego car or not. So I would suggest removing the relative_ego_origin vector.

Comment (SoykaDaimler) [inline osi_object.proto]: What do other people think? Would replacing relative_ego_origin disrupt some OSI logic?
@nHirsenkorn @TimoHankeBMW Your input would be appreciated.
If you agree I could make the necessary changes.

Comment (nHirsenkorn) [inline osi_object.proto]: No, it would not disrupt some logic. If you need it we must add it. However it will delay the initial rollout, as more complexity to the simulator and to the sensor model framework is added. That is the reason why the initial framework has such a minimal number of coordinate systems and reference points.

Comment (SoykaDaimler) [inline osi_object.proto]: OK, so we replace relative_ego_origin vector then and use the two other offset vectors. I will make the adjustments in this PR

Comment (SoykaDaimler) [inline osi_object.proto]: One more question: in which coordinate system was relative_ego_origin defined? In world coordinates? Then it has to be updated each frame. Or in object coordinates (then it would be constant)? That would be consistent with our previous discussion, about the necessity to rotate relative_ego_origin from object to world frame.
In case we have an object frame, we should also define it in the wiki.

Comment (SoykaDaimler) [inline osi_object.proto]: I assumed object coordinates for now.

Comment (nHirsenkorn) [inline osi_object.proto]: object coordinates is correct

File: osi_object.proto [hunk 3]
```diff
@@ -55,13 +51,9 @@ message Vehicle
...
     
-    /// The vector pointing from the bounding box center point (base.position) to the middle (in x, y and z) 
-    /// of the rear axle under neutral load conditions.
-    optional Vector3d bbcenter_to_rear = 11;
-
 	/// The vector pointing from the bounding box center point (base.position) to the middle (in x, y and z) 
-	/// of the front axle under neutral load conditions.    		
-	optional Vector3d bbcenter_to_front = 12;
+	/// of the front axle under neutral load conditions. In object coordinates.    		
+	optional Vector3d bbcenter_to_front = 11;
```
Comment (nHirsenkorn) [inline osi_object.proto]: please correct the identation

Comment (SoykaDaimler) [inline osi_object.proto]: Sorry, there seems to be some problem with tabs and spaces. I corrected it now

**Comment by nHirsenkorn** [comment]
we can merge after a second review.

File: osi_object.proto (hunk 2 above)
Comment (TimoHankeBMW) [inline osi_object.proto]: Replacing relative_ego_origin is fine, especially since it is only a rename and the addition of the front axis coordinate system now.
The (only) purpose of relative_ego_origin is/was to allow the transformation to a frame relative to the rear_axis_coordinate system of a vehicle (specifically ego-vehicles, that's why it wasn't required for all vehicles). That is still possible with the new definitions, so everything is perfectly fine there.

**Comment by SoykaDaimler** [comment]
@nHirsenkorn I thought the versioning is done by the people that merge. Therefore I did not change anything there. So should I do this, or who will change it? Would we then be at 3.0.0?

**Comment by TimoHankeBMW** [comment]
I think the comment for the base_polygon still needs some refinement.
Currently it states: "General definitions: The polygon is defined in the local object frame: x pointing forward and y to the left."
This defines the (x/y-) axes, but not the position of the origin for the purpose of the Vector2d-coordinates. Is "local object frame" relative to the center of the bounding box (i.e. position-coordinates of the vehicle, that would make most sense to me and this meaning is implied as far as I can see) or relative to the front or rear axis?
To avoid confusion, we should specifically include this in the comment (as we now have 3 possible alternatives for the origin of the vehicle frame).

**Comment by SoykaDaimler** [comment]
@TimoHankeBMW 
Thanks for your comments! Yes, I would suggest to use the middle of the bounding box as origin for the object frame. I will add it to the description. 
In general, I think that all coordinate frames should be defined once in a central location instead of defining them over and over again for each field. See my issue #14

**Comment by AndreasVIRES** [comment]
@SoykaDaimler I'm confused... I thought you need world coordinates for the polygon?
```
/// **The polygon is defined in the local object frame**: x pointing forward and y to the left.
/// The origin is the center of the bounding box.
```

**Comment by SoykaDaimler** [comment]
@AndreasVIRES Initially that was my idea, but after our discussion that changed. Therefore I changed the contour to object coordinates. From there I'll calculate the world coordinates myself ;-)

**Comment by nHirsenkorn** [comment]
@SoykaDaimler sounds reasonable to let the merger change the version; moreover it should be regarded as minor change. I added a Wiki article regarding the versioning and added it to the workflow.
